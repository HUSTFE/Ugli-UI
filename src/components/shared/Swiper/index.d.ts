export interface ISwiperOptions {
    startSlideIndex?: number;
    speed?: number;
    auto?: 0;
    continuous?: boolean;
    disableScroll?: false;
    resistance?: number;
    slideDidChange?(slideIndex: number): any;
}
export declare class Swiper {
    private _container;
    private _options;
    private _slides;
    private _wrapper;
    private _slideWidth;
    private _index;
    private _currentTranslate;
    private _touchStartPosition;
    private _touchStartTime;
    private _delta;
    private _isPastingBounds;
    private _mousedown;
    private _isScrolling;
    private _autoInterval;
    constructor(container: HTMLElement | string, options?: ISwiperOptions);
    /**
     * Slide the swiper to specified index.
     * @param index
     * @param callback
     */
    slideTo(index: number, callback?: (currentIndex: number) => void): void;
    /**
     * Slide the swiper to previous index.
     */
    preSlide(): void;
    /**
     * Slide the swiper to next index.
     */
    nextSlide(): void;
    /**
     * Destroy this slider.
     */
    destroy(): void;
    /**
     * Start auto play.
     * @param time
     */
    startAuto(time?: number): void;
    /**
     * Stop auto play.
     */
    stopAuto(): void;
    private _slideToOnce(index, speed, callback);
    private _onTouchStart;
    private _onTouchMove;
    private _onTouchEnd;
    private _onMouseDown;
    private _onMouseMove;
    private _onMouseUp;
    private _onMouseOut;
    private _start(x, y);
    private _move(x, y, e);
    private _end;
    private _translate(index, translateValue, transition?, callback?);
    private _circle(index);
    /**
     * set current slide's index and slide to the specified slide.
     */
    currentIndex: number;
}
export default Swiper;
