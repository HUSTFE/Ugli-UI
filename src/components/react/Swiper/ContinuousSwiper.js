/**
 * +---------------------------------------------------------------------+
 * |                                                                     |
 * | Faraway's Swiper, even worse than a lot similar swipers on github.  |
 * | Write for practice                                                  |
 * | Faraway <hzylovelyl@gmail.com>                                      |
 * |                                                                     |
 * +---------------------------------------------------------------------+
 */
function Swiper(container, options = {}) {
  this.options = {
    ...({
      startSlideIndex: 1,
      speed: 300,
      // auto: 0,
      continuous: true,
      // disableScroll: false,
      // stopPropagation: false,
      resistance: 0.5,
      // direction: Swiper.DIRECTION_VERTICAL,
      callback: () => {},
      transitionEnd: () => {},
    }),
    options,
  }

  // Written for modern browser, we assume that the
  // environment supports transition, transfrom, and
  // addEventListener

  if (!container) {
    throw TypeError('An DOM element must be provided as container to setup swiper!')
  }
  container = (typeof container === 'string'
    ? document.querySelector(container)
    : container
  )
  if (!container || !(container instanceof HTMLElement)) {
    throw TypeError('Invalid selector!')
  }

  // I prefer DOM structure like this:
  //
  // <div class='swipe'>
  //   <div class='swipe-wrap'>
  //     <div></div>
  //     <div></div>
  //     <div></div>
  //   </div>
  // </div>

  // So, we will check if DOM structure matches.
  // Use children instead of childNodes to filter Text node.
  if (container.children.length <= 0) {
    throw DOMError('The container element you provide has no child.')
  }

  // wrapper is clear enough as a VAR's name
  const wrapper = container.children[0]

  // if (wrapper.children.length <= 0) {
  //   throw DOMError('No slides provided.')
  // }

  const slides = wrapper.children

  const slideWidth = container.clientWidth

  // We extend the wrapper's width to `slides.length * slideWidth`
  wrapper.style.width = `${slideWidth * slides.length}px`

  // So, obviously, container must have style `overflow: hidden`.
  container.style.overflow = 'hidden'
  container.style.overflow = 'relative'

  // Then we set all slides to `float: left`.
  // like this:

  //                                      wrapper
  // +---------------+---------------+---------------+----------------+----------------+
  // |               |               |               |                |                |
  // |               |               |               |                |                |
  // |               |               |               |                |                |
  // |    slide 1    |       2       |       3       |       4        |       5        |
  // |               |               |               |                |                |
  // |               |               |               |                |                |
  // |   container   |               |               |                |                |
  // |               |               |               |                |                |
  // |               |               |               |                |                |
  // |               |               |               |                |                |
  // +---------------+---------------+---------------+----------------+----------------+

  // `float: left` is the most important style for each slide.
  // We also need to add `left: -${i * slideWidth}px`

  const currentTranslate = new Array(slides.length).fill(0) // cache the wrapper translate for smooth swipe.
  for (let i = 0; i < slides.length; i += 1) {
    slides[i].style.cssFloat = 'left'
    slides[i].style.position = 'relative'
    slides[i].style.width = `${slideWidth}px`

    slides[i].style.transform = `${-slideWidth}px`
    slides[i].style.left = `-${i * slideWidth}px`

    currentTranslate[i] = -slideWidth
  }

  // But in this way, our wrapper element will be height: 0,
  // so we add `overflow: hidden` to wrapper's style to avoid
  // wrapper element collapsing.
  wrapper.style.overflow = 'hidden'
  wrapper.style.overflow = 'relative'

  // Then we define a helper function to calculate correct index.
  // This function help us calculate next or prev slide's index.
  const circle = index => (slides.length + (index % slides.length)) % slides.length
  // `this.index` will be update during user swiping.
  this.index = circle(this.options.startSlideIndex) || 0
  slides[circle(this.index - 1)].style.transform = `translate3d(${slideWidth}px, 0px, 0px)`
  currentTranslate[circle(this.index - 1)] = slideWidth

  slides[this.index].style.transform = 'translate3d(0px, 0px, 0px)'
  currentTranslate[this.index] = 0

  slides[circle(this.index + 1)].style.transform = `translate3d(${-slideWidth}px, 0px, 0px)`
  currentTranslate[circle(this.index + 1)] = -slideWidth

  // After we add `left: -${i * slideWidth}px`

  //                                      wrapper
  // +---------------+
  // |               |
  // |               |
  // |               |
  // |    slide 5    |
  // |               |
  // |               |
  // |   container   |
  // |               |
  // |               |
  // |               |
  // +---------------+

  // You can see, all other slides have been covered by the last one.
  // We have finished basic layout.

  let touchStartPosition = { x: 0, y: 0 }
  let touchStartTime = 0
  const delta = { x: 0, y: 0 }
  let isPastBounds = false

  // let destroyed = false // if the swiper has been destroyed.

  // now we define some helper functions

  const onTouchStart = (e) => {
    if (e.touches.length === 1) {
      // we record this two values to calculate swipe speed.
      touchStartPosition = {
        x: e.touches[0].pageX,
        y: e.touches[0].pageY,
      }
      touchStartTime = Date.now()
    }
  }

  const onTouchMove = (e) => {
    if (e.touches.length === 1) {
      const touch = e.touches[0]
      delta.x = touch.pageX - touchStartPosition.x
      delta.y = touch.pageY - touchStartPosition.y

      isPastBounds =
        (!this.index && delta.x > 0) || // if first slide and slide amt is greater than 0
        (this.index === slides.length - 1 && delta.x < 0) // or if last slide and slide amt is less than 0

      let deltaX
      if (!isPastBounds || this.options.continuous) {
        // it is not past bounds or it is continuous
        deltaX = delta.x
      } else {
        // Use this function as resistance function:
        // ^ y
        // |                                 y = x^(1-r/2), 0 <= r <= 1.
        // |                                                   XXXX
        // |                                 XXXXX XX XXXX XXXX
        // |                      XXXXXXXXX X
        // |                 XXXXX
        // |             XXXX
        // |          XXX
        // |        XX
        // |      XX
        // |    XX
        // |   XX
        // |  XX
        // | X
        // |X
        // X---------------------------------------------------------> x
        const sign = Math.abs(delta.x) / delta.x
        deltaX = (sign * (Math.abs(delta.x) ** (1 - (this.options.resistance / 2))))
      }
      slides[circle(this.index - 1)].style.transform
        = `translate3d(${deltaX - currentTranslate[circle(this.index - 1)]}px, 0px, 0px)`
      slides[this.index].style.transform
        = `translate3d(${deltaX - currentTranslate[this.index]}px, 0px, 0px)`
      slides[circle(this.index + 1)].style.transform
        = `translate3d(${deltaX - currentTranslate[circle(this.index + 1)]}px, 0px, 0px)`
    }
  }

  const onTouchEnd = (e) => {
    if (e.changedTouches.length === 1) {
      // After a touch is end, we need to know if this
      // touch is a valid swipe
      const duration = Date.now() - touchStartTime

      // this judgement is acceptable to people.
      const isValidSlide =
        duration < 250 && // if slide duration is less than 250ms
          (
            Math.abs(delta.x) > 20 || // and if slide amt is greater than 20px
            Math.abs(delta.x) > slideWidth / 2 // or if slide amt is greater than half the width
          )

      // user accept continuous, so we allow swipe.
      if (this.options.continuous) isPastBounds = false

      // determine direction of swipe (true:right, false:left)
      const direction = delta.x < 0

      if (isValidSlide && !isPastBounds) {
        if (direction) { // right
          this.index = circle(this.index + 1)
        } else {
          this.index = circle(this.index - 1)
        }
      } else {
        this.index = circle(this.index)
      }
      // wrapper.style.transition = `transform ${this.options.speed}ms`
      // wrapper.style.transform = `translate3d(${circle(this.index) * -slideWidth}px, 0px, 0px)`

      // write a function with a name to removeEventListener()
      // const onTransitionEnd = (event) => {
      //   wrapper.style.transition = ''
      //   // call user's callback
      //   this.options.transitionEnd.call(event, this.index, slides[this.index])
      //   wrapper.removeEventListener('transitionend', onTransitionEnd)
      // }
      // wrapper.addEventListener('transitionend', onTransitionEnd)

      // currentTranslate += touchStartPosition.x - touch.pageX
      slides[circle(this.index - 1)].style.transform = `translate3d(${slideWidth}px, 0px, 0px)`
      currentTranslate[circle(this.index - 1)] = slideWidth

      slides[this.index].style.transform = 'translate3d(0px, 0px, 0px)'
      currentTranslate[this.index] = 0

      slides[circle(this.index + 1)].style.transform = `translate3d(${-slideWidth}px, 0px, 0px)`
      currentTranslate[circle(this.index + 1)] = -slideWidth
    }
  }

  // const onMouseDown = (e) => {

  // }

  // const onMouseMove = (e) => {

  // }

  // const onMouseUp = (e) => {

  // }

  wrapper.addEventListener('touchstart', onTouchStart, false)
  wrapper.addEventListener('touchmove', onTouchMove, false)
  wrapper.addEventListener('touchend', onTouchEnd, false)

  // wrapper.addEventListener('mousedown', onMouseDown)
  // wrapper.addEventListener('mousemove', onMouseMove)
  // wrapper.addEventListener('mouseup', onMouseUp)

  Swiper.DIRECTION_VERTICAL = 'vertical'
  Swiper.DIRECTION_HORIZON = 'horizon'

  Swiper.prototype.destroy = function () {
    // destroyed = true

    wrapper.removeEventListener('touchstart', onTouchStart)
    wrapper.removeEventListener('touchmove', onTouchMove)
    wrapper.removeEventListener('touchend', onTouchEnd)
  }

  Swiper.prototype.prevSlide = function () {

  }

  Swiper.prototype.nextSlide = function () {

  }
  // this.wrapper = wrapper
  Swiper.prototype.slideTo = function () {

  }

  Swiper.prototype.update = function () {

  }
}

window.Swiper = Swiper
