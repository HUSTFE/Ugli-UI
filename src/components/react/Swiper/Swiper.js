/**
 * +---------------------------------------------------------------------+
 * | Faraway's Swiper, even worse than a lot similar swipers on github.  |
 * | Write for practice                                                  |
 * | Faraway <hzylovelyl@gmail.com>                                      |
 * +---------------------------------------------------------------------+
 */

export default class Swiper {
  static DIRECTION_VERTICAL = '@@swiper vertical'
  static DIRECTION_HORIZON = '@@swiper horizon'

  constructor(container, options = {
    startSlideIndex: 0,
    speed: 300,
    auto: 0,
    continuous: true,
    disableScroll: false,
    stopPropagation: false,
    debounce: true,
    direction: Swiper.DIRECTION_VERTICAL,
    callback: () => {},
    transitionEnd: () => {},
  }) {
    // Written for modern browser, we assume that the
    // environment supports transition, transfrom, and
    // addEventListener

    if (!container) {
      throw TypeError('An DOM element must be provided as container to setup swiper!')
    }
    container = (typeof container === 'string'
      ? document.querySelector(container)
      : container
    )
    if (!container || !(container instanceof HTMLElement)) {
      throw TypeError('Invalid selector!')
    }

    // I prefer DOM structure like this:
    //
    // <div class='swipe'>
    //   <div class='swipe-wrap'>
    //     <div></div>
    //     <div></div>
    //     <div></div>
    //   </div>
    // </div>

    // So, we will check if DOM structure matches.
    // Use children instead of childNodes to filter Text node.
    if (container.children.length <= 0) {
      throw DOMError('The container element you provide has no child.')
    }

    // wrapper is clear enough as a VAR's name
    const wrapper = container.children[0]

    // if (wrapper.children.length <= 0) {
    //   throw DOMError('No slides provided.')
    // }

    const slides = wrapper.children

    const slideWidth = container.clientWidth

    // We extend the wrapper's width to `slides.length * slideWidth`
    wrapper.style.width = `${slideWidth * slides.length}px`

    // So, obviously, container must have style `overflow: hidden`.
    container.style.overflow = 'hidden'

    // Then we set all slides to `float: left`.
    // like this:

    //                       wrapper
    // +**********---------+---------+---------+---------+
    // *         *         |         |         |         |
    // * slide 1 *    2    |    3    |    4    |    5    |
    // *         *         |         |         |         |
    // *container*         |         |         |         |
    // *         *         |         |         |         |
    // *         *         |         |         |         |
    // ***********---------+---------+---------+---------+

    // `float: left` is the most important style for each slide.
    for (let i = 0; i < slides.length; i += 1) {
      slides[i].style.cssFloat = 'left'
    }

    // But in this way, our wrapper element will be height: 0,
    // so we add `overflow: hidden` to wrapper's style to avoid
    // wrapper element collapsing.
    wrapper.style.overflow = 'hidden'
  }
}
