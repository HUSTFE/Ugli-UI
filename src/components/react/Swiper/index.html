

<html>


<head>
  <meta charset="utf-8">

  <meta name="viewport"
    content="width=device-width, initial-scale=1">
<script src="./ContinuousSwiper.js" type="text/javascript"></script>
  <title>Storybook</title>
</head>
<style>
  #slider {
  width: 300px;
}

.swipe-wrap > div { background-color: red;
  height: 100px;
}
</style>
<body>
<div id='slider' class='swipe'>
  <div class='swipe-wrap'>
    <div>555</div>
    <div>666</div>
    <div>777</div>
  </div>
</div>
<script>
  var a = new Swiper('#slider')
</script>
</body>
</html>
