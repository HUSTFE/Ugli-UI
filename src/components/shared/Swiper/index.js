import combine from './utils/combine';
export class Swiper {
    constructor(container, options = {}) {
        this._container = null;
        this._options = {};
        // will be changed during sliding
        this._index = 0; // current slide index.
        this._touchStartPosition = { x: 0, y: 0 }; // position where slide start, will be update during touch
        this._touchStartTime = 0; // timestamp (in ms)
        this._delta = { x: 0, y: 0 }; // vector of touch delta
        this._isPastingBounds = false; // flag if current swipe is pasting bounds
        // /**
        //  * Slide the swiper to specified index.
        //  * @param index
        //  * @param callback
        //  */
        // public slideTo(index: number, callback: (currentIndex: number) => void) {
        // }
        this._onTouchStart = (e) => {
            if (e.touches.length === 1) {
                // we record this two values to calculate swipe speed.
                this._touchStartPosition = {
                    x: e.touches[0].pageX,
                    y: e.touches[0].pageY,
                };
                this._touchStartTime = Date.now();
            }
        };
        this._onTouchMove = (e) => {
            if (e.touches.length === 1) {
                const touch = e.touches[0];
                const options = this._options;
                // dx will be changed due to resistance.
                let dx = this._delta.x = touch.pageX - this._touchStartPosition.x;
                this._delta.y = touch.pageY - this._touchStartPosition.y;
                this._isPastingBounds =
                    (!this._index && dx > 0) || // if first slide and slide amt is greater than 0
                        (this._index === this._slides.length - 1 && dx < 0); // or if last slide and slide amt is less than 0
                if (this._isPastingBounds && !options.continuous) {
                    // We will make resistance for better xp
                    // Use this function as resistance function:
                    // ^ y
                    // |                                 y = x^(1-r/2), 0 <= r <= 1.
                    // |                                                   XXXX
                    // |                                 XXXXX XX XXXX XXXX
                    // |                      XXXXXXXXX X
                    // |                 XXXXX
                    // |             XXXX
                    // |          XXX
                    // |        XX
                    // |      XX
                    // |    XX
                    // |   XX
                    // |  XX
                    // | X
                    // |X
                    // X---------------------------------------------------------> x
                    const sign = Math.abs(dx) / dx;
                    dx = (sign * (Math.pow(Math.abs(dx), (1 - ((options.resistance || 0) / 2)))));
                }
                const cachedCurrentTranslate = this._currentTranslate;
                const [t1, t2, t3] = [
                    dx + cachedCurrentTranslate[this._circle(this._index - 1)],
                    dx + cachedCurrentTranslate[this._circle(this._index)],
                    dx + cachedCurrentTranslate[this._circle(this._index + 1)],
                ];
                this._translate(this._index, t2);
                // this._translate(this._index - 1, t1)
                // this._translate(this._index + 1, t3)
                if (options.continuous) {
                    // when user want it continuous, we need to translate
                    // both left & right.
                    this._translate(this._index - 1, t1);
                    this._translate(this._index + 1, t3);
                }
                else if (!this._index && dx > 0) {
                    // when user want it not continuous, if it is pasting
                    // left bounds, we won't show the left.
                    this._translate(this._index + 1, t3);
                }
                else if (this._index === this._slides.length - 1 && dx < 0) {
                    // when user want it not continuous, if it is pasting
                    // right bounds, we won't show the right.
                    this._translate(this._index - 1, t1);
                }
                else {
                    this._translate(this._index - 1, t1);
                    this._translate(this._index + 1, t3);
                }
            }
        };
        this._onTouchEnd = (e) => {
            if (e.changedTouches.length === 1) {
                // After a touch is end, we need to know if this
                // touch is a valid swipe.
                const duration = Date.now() - this._touchStartTime;
                // cache for better performance, `let` is used for re-cache
                const dx = this._delta.x;
                const slideWidth = this._slideWidth;
                let cachedIndex = this._index;
                // this judgement is acceptable to most people.
                const isValidSlide = (duration < 250 && // if slide duration is less than 250ms
                    Math.abs(dx) > 20 // and if slide amt is greater than 20px
                ) ||
                    Math.abs(dx) > this._slideWidth / 2; // or if slide amt is greater than half the width
                // user accept continuous, so we allow swipe.
                if (this._options.continuous) {
                    this._isPastingBounds = false;
                }
                // determine direction of swipe (true:right, false:left)
                const direction = dx < 0;
                const speed = this._options.speed || 0;
                if (isValidSlide && !this._isPastingBounds) {
                    if (direction) {
                        cachedIndex = this._index = this._circle(cachedIndex + 1);
                    }
                    else {
                        cachedIndex = this._index = this._circle(cachedIndex - 1);
                    }
                }
                else {
                    cachedIndex = this._index = this._circle(cachedIndex);
                }
                const [i1, i2, i3] = [
                    this._circle(this._index - 1),
                    this._circle(this._index),
                    this._circle(this._index + 1),
                ];
                // This is most important because without this statement,
                // our slides will slide through the wrapper Element, giving
                // our current slide a flash.
                this._translate(this._circle(this._index + 2), this._slideWidth);
                // Then we just need to deal with left, current, right slides.
                this._translate(i1, -slideWidth, speed);
                this._translate(i2, 0, speed);
                this._translate(i3, slideWidth, speed);
                this._currentTranslate[i1] = -slideWidth;
                this._currentTranslate[i2] = 0;
                this._currentTranslate[i3] = slideWidth;
            }
        };
        if (typeof container === 'string') {
            this._container = document.querySelector(container);
        }
        this._container = this._container || null;
        if (!this._container) {
            throw new TypeError('A container is needed as `Element` to create an instance of Swiper.');
        }
        // init the config
        this._options = combine({
            resistance: 0.5,
            speed: 300,
            startSlideIndex: 1,
            // auto: 0,
            continuous: false,
            // disableScroll: false,
            // stopPropagation: false,
            shouldSlideChange: () => true,
            slideDidChange: () => undefined,
        }, options);
        // Written for modern browser, we assume that the
        // environment supports transition, transfrom, and
        // addEventListener
        // I prefer DOM structure like this:
        //
        // <div class='swipe'>
        //   <div class='swipe-wrap'>
        //     <div></div>
        //     <div></div>
        //     <div></div>
        //   </div>
        // </div>
        // So, we will check if DOM structure matches.
        // Use children instead of childNodes to filter Text node.
        if (this._container.children.length <= 0) {
            throw new ReferenceError('The container element you provide has no child.');
        }
        this._wrapper = this._container.children[0];
        this._slides = Array.from(this._wrapper.children);
        this._slideWidth = this._container.clientWidth;
        // We extend the wrapper's width to `slides.length * slideWidth`
        this._wrapper.style.width = `${this._slideWidth * this._slides.length}px`;
        // So, obviously, container must have style `overflow: hidden`.
        this._container.style.overflow = 'hidden';
        this._container.style.overflow = 'relative';
        // Then we set all slides to `float: left`.
        // like this:
        //                                      wrapper
        // +---------------+---------------+---------------+----------------+----------------+
        // |               |               |               |                |                |
        // |               |               |               |                |                |
        // |               |               |               |                |                |
        // |    slide 1    |       2       |       3       |       4        |       5        |
        // |               |               |               |                |                |
        // |               |               |               |                |                |
        // |   container   |               |               |                |                |
        // |               |               |               |                |                |
        // |               |               |               |                |                |
        // |               |               |               |                |                |
        // +---------------+---------------+---------------+----------------+----------------+
        this._currentTranslate = new Array(this._slides.length).fill(0);
        // `float: left` is the most important style for each slide.
        // We also need to add `left: -${i * slideWidth}px`
        this._slides.forEach((slide, i) => {
            slide.style.cssFloat = 'left';
            slide.style.position = 'relative';
            slide.style.width = `${this._slideWidth}px`;
            slide.style.left = `${-i * this._slideWidth}px`;
            slide.style.transform = `translate3d(${this._slideWidth}px, 0px, 0px)`;
            this._currentTranslate[i] = this._slideWidth;
        });
        // But in this way, our wrapper element will be height: 0,
        // so we add `overflow: hidden` to wrapper's style to avoid
        // wrapper element collapsing.
        this._wrapper.style.overflow = 'hidden';
        this._wrapper.style.overflow = 'relative';
        // After we add `left: -${i * slideWidth}px`
        //                                      wrapper
        // +---------------+
        // |               |
        // |               |
        // |               |
        // |    slide 5    |
        // |               |
        // |               |
        // |   container   |
        // |               |
        // |               |
        // |               |
        // +---------------+
        // You can see, all other slides have been covered by the last one.
        // We have finished basic layout.
        this._index = this._circle(this._options.startSlideIndex || 0) || 0;
        this._translate(this._index - 1, -this._slideWidth);
        this._translate(this._index, 0);
        this._translate(this._index + 1, this._slideWidth);
        // slides[circle(this.index - 1)].style.transform = `translate3d(${slideWidth}px, 0px, 0px)`
        this._currentTranslate[this._circle(this._index - 1)] = -this._slideWidth;
        this._currentTranslate[this._index] = 0;
        this._currentTranslate[this._circle(this._index + 1)] = this._slideWidth;
        // we need to add Event listener to make our swiper be live.
        this._wrapper.addEventListener('touchstart', this._onTouchStart, false);
        this._wrapper.addEventListener('touchmove', this._onTouchMove, false);
        this._wrapper.addEventListener('touchend', this._onTouchEnd, false);
    }
    _translate(index, translateValue, transition = 0) {
        const circledIndex = this._circle(index);
        // cache the style reference to make it faster.
        const style = this._slides[circledIndex].style;
        // Only when transition is valid can we set transition & EventListener
        if (transition > 0) {
            style.transition = `transform ${transition}ms`;
            const slideElement = this._slides[circledIndex];
            // const preTransitionValue: string | null = style.transition
            const onTransitionEnd = () => {
                style.transition = '';
                slideElement.removeEventListener('transitionend', onTransitionEnd);
            };
            slideElement.addEventListener('transitionend', onTransitionEnd);
        }
        else {
            if (style.transition) {
                // when `style.transition` has a valid value, we need to
                // remove it, otherwise we will have a transition even though
                // `transition = 0`.
                style.transition = '';
            }
        }
        // use transition3d to accelerate.
        style.transform = `translate3d(${translateValue}px, 0px, 0px)`;
    }
    // Then we define a helper function to calculate correct index.
    // This function help us calculate next or prev slide's index.
    _circle(index) {
        return (this._slides.length + (index % this._slides.length)) % this._slides.length;
    }
    /**
     * set current slide's index and slide to the specified slide.
     */
    get currentIndex() {
        return this._index;
    }
    set currentIndex(index) {
        this._index = index;
    }
}
window.Swiper = Swiper;
export default Swiper;
