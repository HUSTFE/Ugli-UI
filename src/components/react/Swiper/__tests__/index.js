import React from 'react'
import renderer from 'react-test-renderer'
import Swiper from '../index'

describe('Swiper', () => {
  it('Swiper component render without crash', () => {
    const component = renderer.create(<Swiper />)
    const json = component.toJSON()
    expect(json).toMatchSnapshot()
  })
})
