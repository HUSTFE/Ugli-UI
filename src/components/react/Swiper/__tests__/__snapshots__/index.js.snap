// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swiper Swiper component render without crash 1`] = `
<div>
  This is Swiper component.
</div>
`;
